### PR TITLE
Keep the connection alive when try to connect the same user

### LINKED ISSUE #3649: Keep the connection alive when trying to set the same user with an ongoing connection
If a customer, for some reason, tries to connect with the same user during a connection is already created and alive, we keep this previous connection.
It is not considered "perfect integration" on our customer's side, so we log them with a warning and return previous connection data.

### PR DESCRIPTION
**Pull request by JcMinarro**
### 🎯 Goal
Fix #3649
Add an extra conditions that allow to keep a connection alive when try to connect the same users.

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 1]
```diff
@@ -332,16 +332,35 @@ internal constructor(
...
                         "make sure that previously set user is disconnected."
                 )
-                Result.error<ConnectionData>(
-                    ChatError(
-                        "User cannot be set until the previous one is disconnected."
-                    )
-                )
+                when {
+                    userState.user.id != user.id -> {
+                        logger.logE("[setUser] Trying to set different user without disconnect previous one")
```
Comment (filbabic) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:342]: Make sure to end with a full stop (`.`)

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 2]
```diff
@@ -332,16 +332,35 @@ internal constructor(
...
+                            ChatError(
+                                "User cannot be set until the previous one is disconnected."
+                            )
+                        )
+                    }
+                    else -> {
+                        getConnectionId()?.let { Result.success(ConnectionData(userState.user, it)) }
+                            ?: run {
+                                logger.logE(
+                                    "[setUser] Trying to connect the same user twice without a previous completed " +
```
Comment (filbabic) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:353]: I'd change the last part to `without a previously completed connection.`

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 3]
```diff
@@ -332,16 +332,35 @@ internal constructor(
...
+                                        "connection"
+                                )
+                                Result.error(
+                                    ChatError(
+                                        "Failed to connect user. Please check you don't have connected user already."
```
Comment (filbabic) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:358]: A bit of wording `Failed to connect user. Please check you haven't connected a user already.`

File: CHANGELOG.md [hunk 4]
```diff
@@ -19,6 +19,7 @@
 -. `ChatClient::connectUser` as a new optional argument to configure a timeout to be waiting until the connection is established, in another case an error will be returned [#3605](https://github.com/GetStream/stream-chat-android/pull/3605)
 -. `ChatClient::connectAnonymousUser` as a new optional argument to configure a timeout to be waiting until the connection is established, in another case an error will be returned [#3605](https://github.com/GetStream/stream-chat-android/pull/3605)
 -. `ChatClient::connectGuestUser` as a new optional argument to configure a timeout to be waiting until the connection is established, in another case an error will be returned [#3605](https://github.com/GetStream/stream-chat-android/pull/3605)
+-. `ChatClient::connectUser` doesn't return an error on the case there are an previous connection with the same user [#3653](https://github.com/GetStream/stream-chat-android/pull/3653)
```
Comment (kanat) [inline CHANGELOG.md:22]:
```suggestion
-. `ChatClient::connectUser` doesn't return an error in the case there is a previous connection with the same user. [#3653](https://github.com/GetStream/stream-chat-android/pull/3653)
```

**Comment by kanat** [comment]
Hey @JcMinarro,

Overall, it's looking good to me.
I'm wondering if we gonna deprecate the `connectionId` property of `ConectiionData`.

**Comment by JcMinarro** [comment]
@kanat I don't really think we need to deprecate the `connectionId` property from `ConnectionData`, at the end it is a needed property for this entity

**Review by kanat — APPROVED** [review]
LGTM

**Pull request merged**